### ISSUE TITLE
The response-transformer plugin continuously detects outdated configuration to synchronize

### ISSUE THREAD

**Issue by jaygorrell** [Kong/kubernetes-ingress-controller#102]
### Summary

This has specifically been observed with the `response-transformer` plugin but may extend to others as well. 

In these cases, the `KongPlugin` resource is created (and attached to an Ingress), resulting in the entry correctly being added to Kong with a plugin associated to the route. The issue is that by watching the ingress-controller logs, you can see that the plugin is repeatedly trying to be updated.

The larger risk here is that it may trigger things like https://github.com/Kong/kong/pull/3423 this when many plugins are used.

**Kong Ingress controller version**

0.1.0

### Kubernetes version

1.9.3

### Environment

RDS Postgresql Kong Database

### What happened
The `KongPlugin` was created and the Kong configuration was created, but the ingress-controller logs show that it thinks the two are out of sync and tries to persist the changes every 10 minutes.

Immediately after saving a new configuration I can sometimes get a "up to date" message but on the next synchronization check (10 minutes?) it's back to being "outdated".

### Expected behvaior
Updates are not attempted against Kong when the KongPlugin and Kong database appear to be in sync.

### Steps To Reproduce

1. Create a KongPlugin with the following configuration
```
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: add-security-headers
config:
  add:
    headers:
    - "X-Xss-Protection:1; mode=block"
```

2. Attach this plugin to an Ingress using `response-transformer.plugin.konghq.com: add-security-headers`

3. Tail the `ingress-controller` container logs filtered for `add-security-headers`

4. Observe similar to the following logs that repeat regularly
```
I0823 01:27:55.081377       7 kong.go:784] configuring plugin stored in KongPlugin CRD add-security-headers
I0823 01:27:55.133977       7 kong.go:824] plugin add-security-headers configuration in kong is outdated. Updating...
```

### Other Information
I spent a bit of time trying to further isolate this without much success. I did try specifying all components of the response-transformer plugin (remove/append/replace) with empty lists/maps with no success. I'm having trouble determining what exactly is causing it to think the values are mismatched.

**Comment by hbagdi** [comment]
@jaygorrell 
I'm reading the [documentation](https://docs.konghq.com/plugins/response-transformer/) for this plugin and it seems like, it doesn't accept an array of header.

Could you please try the following?
```yaml
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: add-security-headers
config:
  add:
    headers: "X-Xss-Protection:1; mode=block"
```

**Comment by jaygorrell** [comment]
I tried the string formatted values instead of a list but didn't have any luck, but I did get this worked out.

Looking at `pluginDeepEqual` here:
https://github.com/Kong/kubernetes-ingress-controller/blob/b16e4f881ef2415ebb08c99156afaca553e51420/internal/ingress/controller/kong.go#L1177

It iterates over each k/v in the config map, making sure the `KongPlugin` config items are in Kong and with the same value. For most plugins this is straightforward but the transformer plugins have a nested structure that this method doesn't account for that -- so in my case it just compares the `add` key as a whole. Using my example above, in Kong, the structure is this:
```
                "add": {
                    "headers": [
                        "X-Xss-Protection:1; mode=block"
                    ],
                    "json": {}
                },
                "append": {
                    "headers": {},
                    "json": {}
                },
                "remove": {
                    "headers": {},
                    "json": {}
                },
                "replace": {
                    "headers": {},
                    "json": {}
                }
```

The value of the `add` key also contains the empty `json` dict, which causes the check to fail.

Changing my `KongPlugin` to this fixes the issue:
```
apiVersion: configuration.konghq.com/v1
kind: KongPlugin
metadata:
  name: $((basename))-upside-add-security-headers
config:
  add:
    headers:
      - "X-Xss-Protection:1; mode=block"
    json: {}
```

It's also interesting and worth noting that the empty data structure in Kong is a map, while a populated value is a list. That's a little unintuitive but unrelated to Kong Ingress -- aside from how you need to match the values.

I think there's still an issue to be addressed here in code for a better way to compare the objects but if you think this is acceptable, feel free to close the issue.

**Comment by jaygorrell** [comment]
I tried rolling out this temporary fix but hit another snag. Basically the order of the `headers` and `json` keys can vary in how they get persisted in Kong, regardless which order they appear in the `KongPlugin`. In my above example, many of the `json` keys are coming back before the `headers` key. 

This means you would need to create the `KongPlugin`, query Kong to see the order, then adjust the configuration order to match.

**Comment by hbagdi** [comment]
This issue should be fixed now with #106 merged in.
The other issue is with the plugin as noted in https://github.com/Kong/kong/issues/3166.

Closing this, please re-open if needed.
